core/src/main/java/com/linecorp/armeria/server/RedirectService.java

```diff
@@ -0,0 +1,185 @@
+ * to a new location path pattern. Currently only these path patterns are supported:
+ * <ul>
+ *   <li>{@code /new} (no path parameters)</li>
+ *   <li>{@code /new/{var1}} (curly-brace style)</li>
+ *   <li>{@code /new/:var1/new/:var2} (colon style)</li>
+ *   <li>{@code http://host/new} (full url no path parameters)</li>
+ *   <li>{@code http://host/new/{var1}} (full url and curly-brace style param)</li>
+ *   <li>{@code http://host/new/:var1/new/:var2} (full url and colon style param)</li>
+ * </ul>
+ * The {@link RedirectService} will return {@link HttpStatus#TEMPORARY_REDIRECT 307 Temporary Redirect}
+ * by default.
+ * <pre>{@code
+ * ServerBuilder sb = ...;
+ * sb.service("/old/", new RedirectService("/new"));
+ * sb.service("/old/{var}", new RedirectService("/new/{var}");
+ * sb.service("/old/{var}", new RedirectService(ctx -> "/new/" + ctx.pathParam("var"));
+ *
+ * sb.service("/old/{var}", new RedirectService(HttpStatus.MOVED_PERMANENTLY, "/new/{var}");
+ * sb.service("/old/{var}", new RedirectService("http://user:name@localhost:8080/search/{var}");
+ * }</pre>
+ */
+public class RedirectService extends AbstractHttpService {
+
+    private static final Pattern VALID_DEFAULT_URI_PATTERN = Pattern
+            .compile("(?:(?:^https?:/{2}(?:([^:]+:)?[^:@]+@)?[^:]+)(?::[0-9]{1,5})?)?" +
+                    "(?:/[^/{}:]+|/:[^/{}]+|/\\{[^/{}]+})+/?");
+    private static final Pattern PATTERN_PARAMS_START = Pattern.compile("/:|/\\{");
+
+    private final HttpStatus httpStatus;
+    private final Function<? super ServiceRequestContext, String> locationFunction;
+    private Set<String> paramNames;
+
+    /**
+     * Creates a new Redirect {@link Service} that redirects to the location constructed with the specified
+     * {@code locationPattern}.
+     *
+     * @param locationPattern a {@link PathMapping} string that the bound {@link Service} will redirect to.
+     *
+     * @throws IllegalArgumentException if the specified {@code locationPattern} is unsupported or invalid
+     */
+    public RedirectService(String locationPattern) {
+        this(HttpStatus.TEMPORARY_REDIRECT, locationPattern);
+    }
+
+    /**
+     * Creates a new Redirect {@link Service} that redirects to the location returned by
+     * {@code locationFunction}.
+     *
+     * @param locationFunction a {@link Function} that takes a {@link ServiceRequestContext}
+     *      and returns new location path.
+     */
+    public RedirectService(Function<? super ServiceRequestContext, String> locationFunction) {
+        this(HttpStatus.TEMPORARY_REDIRECT, locationFunction);
+    }
+
+    /**
+     * Creates a new Redirect {@link Service} that redirects to the location constructed with the specified
+     * {@code locationPattern}.
+     *
+     * @param redirectStatus the {@link HttpStatus} that the {@link Service} will return.
+     * @param locationPattern a {@link PathMapping} string that the bound {@link Service} will redirect to.
```

Comment:
`a {@link PathMapping} string that ...` -> `the location pattern that is used to generate a redirect location`